flang/lib/Parser/openacc-parsers.cpp

```diff
@@ -221,11 +228,18 @@ TYPE_PARSER(sourced(construct<AccBeginBlockDirective>(
     sourced(Parser<AccBlockDirective>{}), Parser<AccClauseList>{})))
 
 TYPE_PARSER(startAccLine >> sourced(construct<AccEndBlockDirective>("END"_tok >>
-                                sourced(Parser<AccBlockDirective>{}))))
+                                sourced(recovery(Parser<AccBlockDirective>{},
+                                    construct<AccBlockDirective>(pure(
+                                        llvm::acc::Directive::ACCD_data)))))))
 
 TYPE_PARSER(construct<OpenACCBlockConstruct>(
     Parser<AccBeginBlockDirective>{} / endAccLine, block,
-    Parser<AccEndBlockDirective>{} / endAccLine))
+    // TODO: This still allows mismatched directives.
+    recovery(withMessage("expected OpenACC end block directive"_err_en_US,
+                 attempt(Parser<AccEndBlockDirective>{} / endAccLine)),
+        // TODO: Is there a simpler way to build this?
```

Comment:
Please resolve as many of your TODO comments as you can rather than merging them as unresolved problems.

For this one, it might be the case that it would be clearer to move the `pure` call to the top level of the recovery production, i.e. `pure(AccEndBlockDirective{AccBlockDirective{llvm::acc::Directive::ACCD_data}})`.